### PR TITLE
WebCodecs VideoEncoder configure should resolve its promise for unknown codecs

### PR DESCRIPTION
**Pull request by youennf**
#### 7e4ab69bd8f2b884468da81c4ed52a3b86fc2c34
<pre>
WebCodecs VideoEncoder configure should resolve its promise for unknown codecs
<a href="https://bugs.webkit.org/show_bug.cgi?id=260853">https://bugs.webkit.org/show_bug.cgi?id=260853</a>
rdar://problem/114622760

Reviewed by Jean-Yves Avenard.

Update implementation according latest spec to resolve the isConfigureSupported with supported equal to false.
Update implementation according latest spec to close the encoder in case of configuring with an unknown/invalid codec.

We more tightly align when we set m_isMessageQueueBlocked to true and false in the case when configure fails and we close the encoder/decoder.
Although this is not observable, this more closely aligns with the spec steps.

Covered by updated tests.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/video-encoder-config.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/video-encoder-config.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/video-encoder.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/video-encoder.https.any.worker-expected.txt:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoder.cpp:
(WebCore::isValidDecoderConfig):
(WebCore::WebCodecsVideoDecoder::configure):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:
(WebCore::isSupportedEncoderCodec):
(WebCore::isValidEncoderConfig):
(WebCore::WebCodecsVideoEncoder::configure):
(WebCore::WebCodecsVideoEncoder::isConfigSupported):

Canonical link: <a href="https://commits.webkit.org/267668@main">https://commits.webkit.org/267668@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/42830836696c7888042d2114e2881d4b1ba0a867

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19148 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16245 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17828 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18415 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17901 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15095 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19966 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15792 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22458 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16140 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15961 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20285 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16542 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14037 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15688 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4144 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20058 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16378 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp [hunk 1]
```diff
@@ -147,22 +167,16 @@ ExceptionOr<void> WebCodecsVideoEncoder::configure(ScriptExecutionContext& conte
...
-        if (encoderConfig.hasException()) {
-            closeEncoder(Exception { NotSupportedError, encoderConfig.releaseException().message() });
-            return;
-        }
-
         VideoEncoder::create(config.codec, encoderConfig.releaseReturnValue(), [this, weakThis = WeakPtr { *this }](auto&& result) {
             if (!weakThis)
                 return;
 
+            m_isMessageQueueBlocked = false;
```
Comment (jyavenard) [inline Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:174]: this doesn't seem related to the change being described. Can the commit log include a note?

File: Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp [hunk 2]
```diff
@@ -67,10 +67,17 @@ WebCodecsVideoEncoder::~WebCodecsVideoEncoder()
 {
 }
 
-static bool isValidEncoderConfig(const WebCodecsVideoEncoderConfig& config, const Settings::Values& settings)
+static bool isSupportedEncoderCodec(const String& codec, const Settings::Values& settings)
 {
-    // FIXME: Check codec more accurately.
```
Comment (jyavenard) [inline Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:72]: the new test is identical, just reformatted. This FIXME no longer applies?

File: Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp (hunk 1 above)
Comment (youennf) [reply Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:174]: I updated the commit message.

File: Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp (hunk 2 above)
Comment (youennf) [reply Source/WebCore/Modules/webcodecs/WebCodecsVideoEncoder.cpp:72]: Codec specific checks are now resolved asynchronously.
We move them closer to each encoder, which might parse/validate the profile information.
We have done so for decoders already.

**Review by jyavenard — APPROVED** [review]
